printf: fix formatting for zero value when precision is zero

nickorlow opened the pull request:
Fixes issue #7509 

According to the C/C++ reference, "if both the value and precision are 0, the conversion results in no characters. "

This PR also includes code to handle the following exceptions: 
- If width isn't zero, the result should still be padded, but empty
- If the alternate octal output is selected, a `0` should be outputted
- The positive sign should be printed if specified

hz2 commented:
For clippy related issues, it may be helpful to run locally:

```
cargo clippy --all-targets --all-features -- -D warnings
```

nickorlow commented:
Reran lint and addressed issues there. The failing GNU test case was because I didn't switch the precision parser to return a `None` for a negative value instead of `Some(0usize)`.

Should be good to run workflow again.

nickorlow commented:
Updated the one test case. Not sure there's much I can do about the other one that's complaining about linting issues in the `zerocopy` crate.

RenjiSann commented:
Can you squash your commits in a single one for better readability in the main branch ?
Thank you !

nickorlow commented:
@RenjiSann Done! Sorry it took so long to get back on that.

nickorlow commented:
Just rebased on top of main.
